EIPS/eip-signTypedData.md

```diff
@@ -0,0 +1,439 @@
+    'value': 42,
+  },
+];
+const signature = await web3.eth.signTypedData(typedData, signerAddress);
+// or
+const signature = await web3.personal.signTypedData(typedData, signerAddress, '************');
+```
+
+```javascript
+// Signer code JS example
+import * as _ from 'lodash';
+import * as ethAbi from 'ethereumjs-abi';
+
+const schema = _.map(typedData, entry => `${entry.type} ${entry.name}`).join(',');
+// Will generate `string message,uint value` for the above example
+const schemaHash = ethAbi.soliditySHA3(['string'], [schema]);
+const data = _.map(typedData, 'value');
+const types = _.map(typedData, 'type');
+const hash = ethAbi.soliditySHA3(
+  ['bytes32', ...types],
+  [schemaHash, ...data],
+);
+```
+
+
+## Rationale
+
+<!-- The rationale fleshes out the specification by describing what motivated the design and why particular design decisions were made. It should describe alternate designs that were considered and related work, e.g. how the feature is supported in other languages. The rationale may also provide evidence of consensus within the community, and should discuss important objections or concerns raised during discussion. -->
+
+The `encode` function is extended with a new case for the new types. The first byte of the encoding distinguishes the cases. For the same reason it is not safe to start immediately with the domain separator or a `typeHash`. While hard, it may be possible to construct a `typeHash` that also happens to be a prefix of a valid RLP encoded transaction.
+
+The domain separator prevents collision of otherwise identical structures. It is possible that two DApps come up with an identical structure like `Transfer(address from,address to,uint256 amount)` that should not be compatible. By introducing a domain separator the DApp developers are guaranteed that there can be no signature collision.
+
+The domain separator also allows for multiple distinct signatures use-cases on the same struct instance within a given DApp. In the previous example, perhaps signatures from both `from` and `to` are required. By providing two distinct domain separators these signatures can be distinguished from each other.
+
+**Alternative 1**: Use the target contract address as domain separator. This solves the first problem, contracts coming up with identical types, but does not address second use-case. The standard does suggest implementors to use the target contract address where this is appropriate.
+
+The function `hashStruct` starts with a `typeHash` to separate types. By giving different types a different prefix the `encodeData` function only has to be injective for within a given type. It is okay for `encodeData(a)` to equal `encodeData(b)` as long as `typeOf(a)` is not `typeOf(b)`.
+
+### Rationale for `typeHash`
+
+The `typeHash` is designed to turn into a compile time constant in Solidity. For example:
+
+```javascript
+bytes32 constant MESSAGE_TYPEHASH = keccak256(
+  "Message(address from,address to,string contents)");
```

Comment:
I'm really not a fan of including the variable names in the string to be hashed. Variable names don't form part of the interface, and should be entirely up to the implementation to choose. I can't see the example you offer below causing a practical collision, since a contract can't implement both interfaces.